File: 1171-remove-zero-sum-consecutive-nodes-from-linked-list/1171-remove-zero-sum-consecutive-nodes-from-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeZeroSumSublists(ListNode* head) {
        ListNode* dummy = new ListNode(0, head);
        unordered_map<int, ListNode*> Map;
        int s = 0;
        ListNode* cur = dummy;
        while(cur) {
            s += cur->val;
            Map[s] = cur;
            cur = cur->next;
        }
        cur = dummy;
        s = 0;
        while(cur) {
            s += cur->val;
            cur->next = Map[s]->next;
            cur = cur->next;
        }
        
        return dummy->next;
    }
};

File: 1171-remove-zero-sum-consecutive-nodes-from-linked-list/1171-remove-zero-sum-consecutive-nodes-from-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeZeroSumSublists(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0, head)
        prefix_sum_map = {}
        s = 0
        cur = dummy

        # First pass to calculate prefix sums and store nodes in map
        while cur:
            s += cur.val
            prefix_sum_map[s] = cur
            cur = cur.next

        # Second pass to remove nodes part of zero-sum sublists
        cur = dummy
        s = 0
        while cur:
            s += cur.val
            cur.next = prefix_sum_map[s].next
            cur = cur.next

        return dummy.next